types/src/blockchain.rs

```diff
@@ -248,14 +283,37 @@ pub enum BlockchainResponse {
         cumulative_difficulty: u128,
     },
 
+    /// Response to [`BlockchainReadRequest::NextChainEntry`].
+    ///
+    /// If all blocks were unknown `start_height` will be [`None`], the other fields will be meaningless.
+    NextChainEntry {
+        /// The start height of this entry, [`None`] if we could not find the split point.
+        start_height: Option<usize>,
```

Comment:
If `usize` is better here that's ok too, just noting that the `usize // 0 == special_semantics` -> `Option<NonZero<T>>` pattern exists to move the invariant into the type system, also:
```rust
fn main() {
    assert_eq!(size_of::<Option<usize>>(), 16);
    assert_eq!(size_of::<Option<std::num::NonZero<usize>>>(), 8);
}
```
Patch for `Option<NonZero<usize>>`:
```diff
diff --git a/binaries/cuprated/src/p2p/request_handler.rs b/binaries/cuprated/src/p2p/request_handler.rs
index 620c9df..d8fef7f 100644
--- a/binaries/cuprated/src/p2p/request_handler.rs
+++ b/binaries/cuprated/src/p2p/request_handler.rs
@@ -238,7 +238,7 @@ async fn get_chain(
         split_u128_into_low_high_bits(cumulative_difficulty);
 
     Ok(ProtocolResponse::GetChain(ChainResponse {
-        start_height: usize_to_u64(start_height),
+        start_height: usize_to_u64(std::num::NonZero::<usize>::get(start_height)),
         total_height: usize_to_u64(chain_height),
         cumulative_difficulty_low64,
         cumulative_difficulty_top64,
diff --git a/storage/blockchain/src/service/read.rs b/storage/blockchain/src/service/read.rs
index f3e0a65..0263f42 100644
--- a/storage/blockchain/src/service/read.rs
+++ b/storage/blockchain/src/service/read.rs
@@ -634,7 +634,7 @@ fn next_chain_entry(
     };
 
     Ok(BlockchainResponse::NextChainEntry {
-        start_height: Some(first_known_height),
+        start_height: std::num::NonZero::<usize>::new(first_known_height),
         chain_height,
         block_ids,
         block_weights,
diff --git a/types/src/blockchain.rs b/types/src/blockchain.rs
index 58a23b2..7518935 100644
--- a/types/src/blockchain.rs
+++ b/types/src/blockchain.rs
@@ -288,7 +288,7 @@ pub enum BlockchainResponse {
     /// If all blocks were unknown `start_height` will be [`None`], the other fields will be meaningless.
     NextChainEntry {
         /// The start height of this entry, [`None`] if we could not find the split point.
-        start_height: Option<usize>,
+        start_height: Option<std::num::NonZero<usize>>,
         /// The current chain height.
         chain_height: usize,
         /// The next block hashes in the entry.
```
Patch for `usize`:
```diff
diff --git a/binaries/cuprated/src/p2p/request_handler.rs b/binaries/cuprated/src/p2p/request_handler.rs
index 620c9df..adc9be1 100644
--- a/binaries/cuprated/src/p2p/request_handler.rs
+++ b/binaries/cuprated/src/p2p/request_handler.rs
@@ -230,9 +230,9 @@ async fn get_chain(
         unreachable!();
     };
 
-    let Some(start_height) = start_height else {
+    if start_height == 0 {
         anyhow::bail!("The peers chain has a different genesis block than ours.");
-    };
+    }
 
     let (cumulative_difficulty_low64, cumulative_difficulty_top64) =
         split_u128_into_low_high_bits(cumulative_difficulty);
diff --git a/storage/blockchain/src/service/read.rs b/storage/blockchain/src/service/read.rs
index f3e0a65..f14e50e 100644
--- a/storage/blockchain/src/service/read.rs
+++ b/storage/blockchain/src/service/read.rs
@@ -600,7 +600,7 @@ fn next_chain_entry(
     // This will happen if we have a different genesis block.
     if idx == block_ids.len() {
         return Ok(BlockchainResponse::NextChainEntry {
-            start_height: None,
+            start_height: 0,
             chain_height: 0,
             block_ids: vec![],
             block_weights: vec![],
@@ -634,7 +634,7 @@ fn next_chain_entry(
     };
 
     Ok(BlockchainResponse::NextChainEntry {
-        start_height: Some(first_known_height),
+        start_height: first_known_height,
         chain_height,
         block_ids,
         block_weights,
diff --git a/types/src/blockchain.rs b/types/src/blockchain.rs
index 58a23b2..e1a6dc9 100644
--- a/types/src/blockchain.rs
+++ b/types/src/blockchain.rs
@@ -285,10 +285,10 @@ pub enum BlockchainResponse {
 
     /// Response to [`BlockchainReadRequest::NextChainEntry`].
     ///
-    /// If all blocks were unknown `start_height` will be [`None`], the other fields will be meaningless.
+    /// If all blocks were unknown `start_height` will be `0`, the other fields will be meaningless.
     NextChainEntry {
-        /// The start height of this entry, [`None`] if we could not find the split point.
-        start_height: Option<usize>,
+        /// The start height of this entry, `0` if we could not find the split point.
+        start_height: usize,
         /// The current chain height.
         chain_height: usize,
         /// The next block hashes in the entry.
```